model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java

```diff
@@ -0,0 +1,308 @@
+import org.keycloak.models.map.common.Serialization;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.MapStorage;
+
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static org.keycloak.common.util.StackUtil.getShortStackTrace;
+
+public class MapGroupProvider implements GroupProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapGroupProvider.class);
+    private static final Predicate<MapGroupEntity> ALWAYS_FALSE = c -> { return false; };
+    private final KeycloakSession session;
+    final MapKeycloakTransaction<UUID, MapGroupEntity> tx;
+    private final MapStorage<UUID, MapGroupEntity> groupStore;
+
+    public MapGroupProvider(KeycloakSession session, MapStorage<UUID, MapGroupEntity> groupStore) {
+        this.session = session;
+        this.groupStore = groupStore;
+        this.tx = new MapKeycloakTransaction<>(groupStore);
+        session.getTransactionManager().enlist(tx);
+    }
+
+    private MapGroupEntity registerEntityForChanges(MapGroupEntity origEntity) {
+        final MapGroupEntity res = Serialization.from(origEntity);
+        tx.putIfChanged(origEntity.getId(), res, MapGroupEntity::isUpdated);
+        return res;
+    }
+
+    private Function<MapGroupEntity, GroupModel> entityToAdapterFunc(RealmModel realm) {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+        return origEntity -> new MapGroupAdapter(session, realm, registerEntityForChanges(origEntity));
+    }
+
+    private Predicate<MapGroupEntity> entityRealmFilter(RealmModel realm) {
+        if (realm == null || realm.getId() == null) {
+            return MapGroupProvider.ALWAYS_FALSE;
+        }
+        String realmId = realm.getId();
+        return entity -> Objects.equals(realmId, entity.getRealmId());
+    }
+
+    @Override
+    public GroupModel getGroupById(RealmModel realm, String id) {
+        if (id == null) {
+            return null;
+        }
+
+        LOG.tracef("getGroupById(%s, %s)%s", realm, id, getShortStackTrace());
+
+
+        UUID uid;
+        try {
+            uid = UUID.fromString(id);
+        } catch (IllegalArgumentException ex) {
+            return null;
+        }
+        
+        MapGroupEntity entity = tx.get(uid, groupStore::get);
+        return (entity == null || ! entityRealmFilter(realm).test(entity))
+                ? null
+                : entityToAdapterFunc(realm).apply(entity);
+    }
+
+    private Stream<MapGroupEntity> getNotRemovedUpdatedGroupsStream() {
+        Stream<MapGroupEntity> updatedAndNotRemovedGroupsStream = groupStore.entrySet().stream()
+                .map(tx::getUpdated)    // If the client has been removed, tx.get will return null, otherwise it will return me.getValue()
```

Comment:
nit:
```suggestion
                .map(tx::getUpdated)    // If the group has been removed, tx.get will return null, otherwise it will return me.getValue()
```